plugins/woocommerce/src/Blocks/Domain/Services/CheckoutLink.php

```diff
@@ -0,0 +1,128 @@
+<?php
+/**
+ * Functionality that takes a static URL, constructs a cart, and redirects to the checkout with a cart session.
+ */
+
+declare(strict_types=1);
+
+namespace Automattic\WooCommerce\Blocks\Domain\Services;
+
+use Automattic\WooCommerce\Blocks\Utils\CartCheckoutUtils;
+use Automattic\WooCommerce\StoreApi\Utilities\CartTokenUtils;
+use Automattic\WooCommerce\StoreApi\Utilities\CartController;
+
+defined( 'ABSPATH' ) || exit;
+
+/**
+ * Checkout Link class.
+ */
+class CheckoutLink {
+	/**
+	 * Initialize the checkout link service.
+	 */
+	public function init() {
+		add_action( 'init', array( $this, 'add_checkout_link_endpoint' ) );
+		add_filter( 'query_vars', array( $this, 'add_query_vars' ), 0 );
+		add_action( 'template_redirect', array( $this, 'handle_checkout_link_endpoint' ) );
+	}
+
+	/**
+	 * Add the checkout link endpoint.
+	 */
+	public function add_checkout_link_endpoint() {
+		add_rewrite_rule( '^checkout-link$', 'index.php?checkout-link=true', 'top' );
+	}
+
+	/**
+	 * Add the checkout link query var.
+	 *
+	 * @param array $vars The query vars.
+	 * @return array The query vars.
+	 */
+	public function add_query_vars( $vars ) {
+		$vars[] = 'checkout-link';
+		return $vars;
+	}
+
+	/**
+	 * Handle the checkout link endpoint.
+	 *
+	 * @return void
+	 */
+	public function handle_checkout_link_endpoint() {
+		if ( ! get_query_var( 'checkout-link' ) ) {
+			return;
+		}
+		wp_safe_redirect( $this->get_checkout_link() );
+		exit;
+	}
+
+	/**
+	 * Process the query params and return the checkout link to redirect to complete with session token.
+	 *
+	 * @return string The checkout link.
+	 */
+	protected function get_checkout_link() {
+		$controller = new CartController();
+		$controller->empty_cart();
+
+		// Populate cart with products.
+		$products = array_filter( explode( ',', wc_clean( wp_unslash( $_GET['products'] ?? '' ) ) ) ); // phpcs:ignore WordPress.Security.NonceVerification.Recommended
+
+		foreach ( $products as $product_id_qty ) {
+			if ( strpos( $product_id_qty, ':' ) !== false ) {
+				list( $product_id, $qty ) = explode( ':', $product_id_qty );
+			} else {
+				$product_id = $product_id_qty;
+				$qty        = 1;
+			}
+			if ( ! absint( $product_id ) ) {
+				continue;
+			}
+			try {
+				$controller->add_to_cart(
```

Comment:
Correct me if I'm wrong but this should just work with variation product ids right? I will test regardless.

Reply:
It should work with anything that does not have special options. You'll get notices if you try to add a variable product. Not sure if we want to add a more deliberate check here.